File: recyclerlibrary/src/main/java/com/cz/recyclerlibrary/adapter/SelectAdapter.java
package com.cz.recyclerlibrary.adapter;

import android.support.v7.widget.RecyclerView;
import android.view.View;

import com.cz.recyclerlibrary.PullToRefreshRecyclerView;
import com.cz.recyclerlibrary.callback.Selectable;

import java.util.ArrayList;
import java.util.List;

import kotlin.ranges.IntRange;

/**
 * Created by cz on 4/3/16.
 * 一个可设置选择模式的数据乱配器
 */
public class SelectAdapter extends RefreshAdapter {
    public static final int MAX_COUNT=Integer.MAX_VALUE;
    public static final int INVALID_POSITION=-1;
    // 三种选择状态
    public static final int CLICK = 0;//单击
    public static final int SINGLE_SELECT = 1;//单选
    public static final int MULTI_SELECT = 2;//多选
    public static final int RECTANGLE_SELECT = 3;//块选择
    private final ArrayList<Integer> multiSelectItems;//选中集
    private PullToRefreshRecyclerView.OnSingleSelectListener singleSelectListener;
    private PullToRefreshRecyclerView.OnMultiSelectListener multiSelectListener;
    private PullToRefreshRecyclerView.OnRectangleSelectListener rectangleSelectListener;
    private int selectPosition;// 选中位置
    private int selectMaxCount;
    private int start, end;//截选范围
    private int mode;//选择模式

    public SelectAdapter(RecyclerView.Adapter adapter) {
        super(adapter);
        selectPosition=-1;
        multiSelectItems = new ArrayList<>();
    }

    /*
     * 设置选择模式
     *
     * @param mode
     */
    public void setSelectMode(int mode) {
        int headersCount = getHeaderViewCount();
        switch (this.mode) {
            case SINGLE_SELECT:
                //清除单选择状态
                int lastSelectPosition=selectPosition;
                selectPosition = -1;
                if(INVALID_POSITION!=lastSelectPosition){
                    notifyItemChanged(lastSelectPosition + headersCount);
                }
                break;
            case MULTI_SELECT:
                List<Integer> lastItems=new ArrayList<>(multiSelectItems);
                multiSelectItems.clear();
                for (Integer position : lastItems) {
                    notifyItemChanged(position+headersCount);
                }
                break;
            case RECTANGLE_SELECT:
                int s = Math.min(start, end) + headersCount, e = Math.max(start, end) + headersCount;
                start = end = 0;
                notifyItemRangeChanged(s, e - s+1);
                break;
        }
        this.mode = mode;
    }

    public void setSelectMaxCount(int count) {
        this.selectMaxCount=count;
    }

    public void setSingleSelectPosition(int position){
        this.selectPosition=position;
        int lastPosition=selectPosition;
        if(0<=position&&position<getItemCount()){
            //如果在正常范围内
            notifyItemChanged(position);
        } else {
            //删掉上一个选择的条目
            notifyItemChanged(lastPosition);
        }
    }

    public int getSingleSelectPosition(){
        return this.selectPosition;
    }


    public void setMultiSelectItems(List<Integer> items){
        multiSelectItems.clear();
        multiSelectItems.addAll(items);
        notifyDataSetChanged();
    }

    public List<Integer> getMultiSelectItems(){
        return multiSelectItems;
    }

    public void setRectangleSelectPosition(int start,int end){
        this.start=start;
        this.end=end;
        notifyItemRangeChanged(start,end-start);
    }

    public IntRange getRectangleSelectPosition(){
        return new IntRange(start,end);
    }

    @Override
    public void onBindViewHolder(RecyclerView.ViewHolder holder, int position) {
        super.onBindViewHolder(holder, position);
        int headersCount = getHeaderViewCount();
        int footerViewCount = getFooterViewCount();
        switch (mode) {
            case SINGLE_SELECT:
                selectPosition(holder,position,headersCount,footerViewCount,selectPosition + headersCount == position);
                break;
            case MULTI_SELECT:
                selectPosition(holder,position,headersCount,footerViewCount,multiSelectItems.contains(position-headersCount));
                break;
            case RECTANGLE_SELECT:
                int s = Math.min(start + headersCount, end + headersCount);
                int e = Math.max(start + headersCount, end + headersCount);
                selectPosition(holder,position,headersCount,footerViewCount,s <= position && e >= position);
                break;
            default:
                selectPosition(holder,position,headersCount,footerViewCount,false);
        }
    }

    private void selectPosition(RecyclerView.ViewHolder holder,int position,int headerCount,int footerCount,boolean select){
        int itemCount = getItemCount();
        if(null!=adapter&&adapter instanceof Selectable&&position>=headerCount&&position<itemCount-footerCount){
            Selectable selectable=(Selectable)adapter;
            selectable.onSelectItem(holder,position-headerCount,select);
        }
    }


    @Override
    protected boolean onItemClick(View v, int position) {
        super.onItemClick(v, position);
        int headersCount = getHeaderViewCount();
        switch (mode) {
            case MULTI_SELECT:
                int lastSize=multiSelectItems.size();
                selectPosition = start = end = -1;
                if (multiSelectItems.contains(position)) {
                    lastSize--;
                    multiSelectItems.remove(Integer.valueOf(position));
                    notifyItemChanged(position + headersCount);
                } else if(multiSelectItems.size()<selectMaxCount){
                    multiSelectItems.add(Integer.valueOf(position));
                    notifyItemChanged(position + headersCount);
                }
                if (null != multiSelectListener) {
                    multiSelectListener.onMultiSelect(v, multiSelectItems,lastSize,selectMaxCount);
                }
                break;
            case RECTANGLE_SELECT:
                if (-1 != start && -1 != end) {
                    int s = start, e = end;
                    start = end = -1;//重置
                    notifyItemRangeChanged(Math.min(s + headersCount, e + headersCount), Math.max(s + headersCount, e + headersCount));
                } else if (-1 == start) {
                    start = position;
                    notifyItemChanged(start + headersCount);
                } else if (-1 == end) {
                    end = position;
                    if (null != rectangleSelectListener) {
                        rectangleSelectListener.onRectangleSelect(start, end);
                    }
                    notifyItemRangeChanged(Math.min(start + headersCount, end + headersCount), Math.max(start + headersCount, end + headersCount));
                }
                break;
            case SINGLE_SELECT:
                start = end = -1;
                multiSelectItems.clear();
                int last=selectPosition;
                selectPosition = position;
                if (null != singleSelectListener) {
                    singleSelectListener.onSingleSelect(v, position, last);
                }
                if(0<=selectPosition&&INVALID_POSITION!=last){
                    notifyItemChanged(last + headersCount);//通知上一个取消
                }
                notifyItemChanged(position + headersCount);//本次选中
                break;
        }
        return CLICK==mode;
    }

    /*
     * 设置单选选择监听
     *
     * @param singleSelectListener
     */
    public void setOnSingleSelectListener(PullToRefreshRecyclerView.OnSingleSelectListener singleSelectListener) {
        this.singleSelectListener = singleSelectListener;
    }

    /*
    * 设置多选选择监听
    *
    * @param singleSelectListener
    */
    public void setOnMultiSelectListener(PullToRefreshRecyclerView.OnMultiSelectListener multiSelectListener) {
        this.multiSelectListener = multiSelectListener;
    }

    /*
    * 设置截取选择监听
    *
    * @param singleSelectListener
    */
    public void setOnRectangleSelectListener(PullToRefreshRecyclerView.OnRectangleSelectListener rectangleSelectListener) {
        this.rectangleSelectListener = rectangleSelectListener;
    }



}

File: recyclerlibrary/src/main/java/com/cz/recyclerlibrary/adapter/SelectAdapter.kt
package com.cz.recyclerlibrary.adapter

import android.support.v7.widget.RecyclerView
import android.view.View

import com.cz.recyclerlibrary.PullToRefreshRecyclerView
import com.cz.recyclerlibrary.callback.Selectable

import java.util.ArrayList

/**
 * Created by cz on 4/3/16.
 * 一个可设置选择模式的数据乱配器
 */
open class SelectAdapter(adapter: RecyclerView.Adapter<RecyclerView.ViewHolder>?) : RefreshAdapter(adapter) {
    companion object {
        val MAX_COUNT = Integer.MAX_VALUE
        val INVALID_POSITION = -1
        // 三种选择状态
        val CLICK = 0//单击
        val SINGLE_SELECT = 1//单选
        val MULTI_SELECT = 2//多选
        val RECTANGLE_SELECT = 3//块选择
    }

    private var singleSelectListener: PullToRefreshRecyclerView.OnSingleSelectListener? = null
    private var multiSelectListener: PullToRefreshRecyclerView.OnMultiSelectListener? = null
    private var rectangleSelectListener: PullToRefreshRecyclerView.OnRectangleSelectListener? = null
    private var selectPosition: Int = -1// 选中位置
    private var selectMaxCount: Int = 0
    private var start: Int = INVALID_POSITION
    private var end: Int = INVALID_POSITION//截选范围
    private var mode: Int = 0//选择模式
    var multiSelectItems = ArrayList<Int>()//选中集
        set(value) {
            field.clear()
            field.addAll(value)
            notifyDataSetChanged()
        }

    /*
     * 设置选择模式
     *
     * @param newMode
     */
    fun setSelectMode(newMode: Int) {
        //清空所有其他状态
        when (this.mode) {
            SINGLE_SELECT -> {
                //清除单选择状态
                val lastSelectPosition = selectPosition
                selectPosition = INVALID_POSITION
                if (INVALID_POSITION != lastSelectPosition) {
                    notifyItemChanged(lastSelectPosition + headerViewCount)
                }
            }
            MULTI_SELECT -> {
                multiSelectItems.let {
                    it.forEach { notifyItemChanged(it + headerViewCount) }
                    it.clear()
                }
            }
            RECTANGLE_SELECT -> {
                val s = Math.min(start, end) + headerViewCount
                val e = Math.max(start, end) + headerViewCount
                notifyItemRangeChanged(s, e - s + 1)
                start = 0
                end = 0
            }
        }
        start = INVALID_POSITION
        end = INVALID_POSITION
        selectPosition=INVALID_POSITION
        multiSelectItems.clear()
        this.mode = newMode
    }

    fun setSelectMaxCount(count: Int) {
        this.selectMaxCount = count
    }

    //如果在正常范围内
    //删掉上一个选择的条目
    var singleSelectPosition: Int
        get() = this.selectPosition
        set(position) {
            val lastPosition = selectPosition
            this.selectPosition = position
            if(lastPosition in 0..(itemCount-1)){
                notifyItemChanged(lastPosition)
            }
            if (position in 0..(itemCount - 1)) {
                notifyItemChanged(position)
            }
        }

    val rectangleSelectPosition: IntRange
        get() = IntRange(start, end)

    fun setRectangleSelectPosition(start: Int, end: Int) {
        this.start = start
        this.end = end
        notifyItemRangeChanged(start, end - start)
    }


    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        super.onBindViewHolder(holder, position)
        val headersCount = headerViewCount
        val footerViewCount = footerViewCount
        when (mode) {
            SINGLE_SELECT -> selectPosition(holder, position, headersCount, footerViewCount, selectPosition + headersCount == position)
            MULTI_SELECT -> selectPosition(holder, position, headersCount, footerViewCount, multiSelectItems.contains(position - headersCount))
            RECTANGLE_SELECT -> {
                val s = Math.min(start + headersCount, end + headersCount)
                val e = Math.max(start + headersCount, end + headersCount)
                selectPosition(holder, position, headersCount, footerViewCount, s <= position && e >= position)
            }
            else -> selectPosition(holder, position, headersCount, footerViewCount, false)
        }
    }

    protected open fun selectPosition(holder: RecyclerView.ViewHolder, position: Int, headerCount: Int, footerCount: Int, select: Boolean) {
        val itemCount = itemCount
        if (null != adapter && adapter is Selectable<*> && position >= headerCount && position < itemCount - footerCount) {
            val selectable = adapter as Selectable<RecyclerView.ViewHolder>
            selectable.onSelectItem(holder, position - headerCount, select)
        }
    }


    override fun onItemClick(v: View, position: Int,realPosition:Int): Boolean {
        super.onItemClick(v, position,realPosition)
        when (mode) {
            MULTI_SELECT -> {
                var lastSize = multiSelectItems.size
                if (multiSelectItems.contains(realPosition)) {
                    lastSize--
                    multiSelectItems.remove(Integer.valueOf(realPosition))
                    notifyItemChanged(position)
                } else if (multiSelectItems.size < selectMaxCount) {
                    multiSelectItems.add(Integer.valueOf(realPosition))
                    notifyItemChanged(position )
                }
                multiSelectListener?.onMultiSelect(v, multiSelectItems, lastSize, selectMaxCount)
            }
            RECTANGLE_SELECT -> if (INVALID_POSITION != start && INVALID_POSITION != end) {
                notifyItemRangeChanged(Math.min(getAdapterPosition(start), getAdapterPosition(end)), Math.max(getAdapterPosition(start), getAdapterPosition(end)))
                end = INVALID_POSITION
                start = end//重置
            } else if (INVALID_POSITION == start) {
                start = realPosition
                notifyItemChanged(position)
            } else if (INVALID_POSITION == end) {
                end = realPosition
                rectangleSelectListener?.onRectangleSelect(start, end)
                notifyItemRangeChanged(Math.min(getAdapterPosition(start), getAdapterPosition(end)), Math.max(getAdapterPosition(start), getAdapterPosition(end)))
            }
            SINGLE_SELECT -> {
                val last = selectPosition
                selectPosition = realPosition
                singleSelectListener?.onSingleSelect(v, realPosition, last)
                if (0 <= selectPosition && INVALID_POSITION != last) {
                    notifyItemChanged(getAdapterPosition(last))//通知上一个取消
                }
                notifyItemChanged(position)//本次选中
            }
        }
        return CLICK == mode
    }

    /*
     * 设置单选选择监听
     *
     * @param singleSelectListener
     */
    fun setOnSingleSelectListener(singleSelectListener: PullToRefreshRecyclerView.OnSingleSelectListener) {
        this.singleSelectListener = singleSelectListener
    }

    /*
    * 设置多选选择监听
    *
    * @param singleSelectListener
    */
    fun setOnMultiSelectListener(multiSelectListener: PullToRefreshRecyclerView.OnMultiSelectListener) {
        this.multiSelectListener = multiSelectListener
    }

    /*
    * 设置截取选择监听
    *
    * @param singleSelectListener
    */
    fun setOnRectangleSelectListener(rectangleSelectListener: PullToRefreshRecyclerView.OnRectangleSelectListener) {
        this.rectangleSelectListener = rectangleSelectListener
    }

    /**
     * 列表单选
     */
    fun onSingleSelect(listener:(View,Int,Int)->Unit){
        this.singleSelectListener=object :PullToRefreshRecyclerView.OnSingleSelectListener{
            override fun onSingleSelect(v: View, newPosition: Int, oldPosition: Int) {
                listener(v,newPosition,oldPosition)
            }
        }
    }

    /**
     * 列表多选事件
     */
    fun onMultiSelect(listener:(View,ArrayList<Int>, Int,Int)->Unit) {
        this.multiSelectListener=object : PullToRefreshRecyclerView.OnMultiSelectListener {
            override fun onMultiSelect(v: View, selectPositions: ArrayList<Int>, lastSelectCount: Int, maxCount: Int) {
                listener(v, selectPositions, lastSelectCount, maxCount)
            }
        }
    }

    /**
     * 块选选中事件
     */
    fun onRectangleSelect(listener:(Int,Int)->Unit){
        this.rectangleSelectListener=object :PullToRefreshRecyclerView.OnRectangleSelectListener{
            override fun onRectangleSelect(startPosition: Int, endPosition: Int) {
                listener(startPosition,endPosition)
            }
        }
    }

}
